socketmode/socketmode_eventhandler.go

```diff
@@ -0,0 +1,218 @@
+		InteractionBlockActionEventMap: interactionBlockActionEventMap,
+		SlashCommandMap:                slackCommandMap,
+		Default: func(e *Event, c *Client) {
+			log.Printf("Unexpected event type received: %v\n", e.Type)
+		},
+	}
+}
+
+// Register a middleare function to use to handle an Event (from socketmode)
+func (r *SocketmodeHandler) Handle(et EventType, f SocketmodeHandlerFunc) {
+	r.EventMap[et] = append(r.EventMap[et], f)
+}
+
+// Register a middleare function to use to handle an Interaction
+func (r *SocketmodeHandler) HandleInteraction(et slack.InteractionType, f SocketmodeHandlerFunc) {
+	r.InteractionEventMap[et] = append(r.InteractionEventMap[et], f)
+}
+
+// Register a middleare function to use to handle an Interaction Block Action referenced by its ActionID
+func (r *SocketmodeHandler) HandleInteractionBlockAction(action_id string, f SocketmodeHandlerFunc) {
+	r.InteractionBlockActionEventMap[action_id] = append(r.InteractionBlockActionEventMap[action_id], f)
+}
+
+// Register a middleare function to use to handle an Event (from slackevents)
+func (r *SocketmodeHandler) HandleEventsAPI(et slackevents.EventAPIType, f SocketmodeHandlerFunc) {
+	r.EventApiMap[et] = append(r.EventApiMap[et], f)
+}
+
+func (r *SocketmodeHandler) HandleSlashCommand(command string, f SocketmodeHandlerFunc) {
+	r.SlashCommandMap[command] = append(r.SlashCommandMap[command], f)
+}
+
+// Register a middleare function to use as a last resort
+func (r *SocketmodeHandler) HandleDefault(f SocketmodeHandlerFunc) {
+	r.Default = f
+}
+
+// RunSlackEventLoop receives the event via the socket
+func (r *SocketmodeHandler) RunEventLoop() {
+
+	go r.runEventLoop()
+
+	r.Client.Run()
+}
+
+func (r *SocketmodeHandler) runEventLoop() {
+	for evt := range r.Client.Events {
+		r.dispatcher(evt)
+	}
+}
+
+func (r *SocketmodeHandler) dispatcher(evt Event) {
+	var ishandled bool
+
+	// Some eventType can be further decomposed
+	switch evt.Type {
+	case EventTypeInteractive:
+		ishandled = r.interactionDispatcher(&evt)
+	case EventTypeEventsAPI:
+		ishandled = r.eventAPIDispatcher(&evt)
+	case EventTypeSlashCommand:
+		ishandled = r.slashCommandDispatcher(&evt)
+	default:
+		ishandled = r.socketmodeDispatcher(&evt)
+	}
+
+	if !ishandled {
+		go r.Default(&evt, r.Client)
+	}
+}
+
+// Dispatch socketmode events to the registered middleware
+func (r *SocketmodeHandler) socketmodeDispatcher(evt *Event) bool {
+	if handlers, ok := r.EventMap[evt.Type]; ok {
+		// If we registered an event
+		for _, f := range handlers {
+			go f(evt, r.Client)
+		}
+
+		return true
+	}
+
+	return false
+}
+
+// Dispatch interactions to the registered middleware
+func (r *SocketmodeHandler) interactionDispatcher(evt *Event) bool {
+	var ishandled bool = false
+
+	interaction, ok := evt.Data.(slack.InteractionCallback)
+	if !ok {
+		fmt.Printf("Ignored %+v\n", evt)
+		return false
+	}
+
+	// Level 1 - socketmode EventType
+	ishandled = r.socketmodeDispatcher(evt)
```

Comment:
I want to be able to handle multiple levels: level 1 is the most generic and level 3 is the most user-friendly.

lvl 3 would be slash commands, shortcuts, workflow steps, etc.

I don't believe that I can create an exhaustive list handlers so that way if someone is stuck they can write middleware that deals with whatever comes in instead of opening an issue saying their use case is not possible.